contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,660 @@
+                self.env(),
+                Event::SwapFeeSet(SwapFeeSet { caller, swap_fee }),
+            );
+
+            self.swap_fee = swap_fee;
+            Ok(())
+        }
+
+        /// Returns current value of the swap_fee parameter
+        #[ink(message)]
+        pub fn swap_fee(&mut self) -> Balance {
+            self.swap_fee
+        }
+
+        /// Sets access_control to a new contract address
+        ///
+        /// Potentially very destructive, can only be called by the contract's Owner.        
+        #[ink(message)]
+        pub fn set_access_control(&mut self, access_control: AccountId) -> Result<(), DexError>
+        where
+            Self: AccessControlled,
+        {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::Owner(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            self.access_control = access_control;
+            Ok(())
+        }
+
+        /// Returns current address of the AccessControl contract that holds the account priviledges for this DEX
+        #[ink(message)]
+        pub fn access_control(&self) -> AccountId {
+            self.access_control
+        }
+
+        /// Terminates the contract.
+        ///
+        /// Can only be called by the contract's Owner.
+        #[ink(message)]
+        pub fn terminate(&mut self) -> Result<(), DexError> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::Owner(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            self.env().terminate_contract(caller)
+        }
+
+        /// Returns own code hash
+        #[ink(message)]
+        pub fn code_hash(&self) -> Result<Hash, DexError> {
+            self.env()
+                .own_code_hash()
+                .map_err(|why| DexError::InkEnv(format!("Can't retrieve own code hash: {:?}", why)))
+        }
+
+        fn new_init(&mut self, ubik: AccountId, cyberiad: AccountId, lono: AccountId) {
+            self.access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+            self.ubik = ubik;
+            self.cyberiad = cyberiad;
+            self.lono = lono;
+        }
+
+        /// Transfers a given amount of a PSP22 token to a specified using the callers own balance
+        fn transfer_tx(
+            &self,
+            token: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
```

Comment:
I'm afraid, that you need at least think about reentrancy here - see my short report on Slack

in brief - contract `A` can call this dex and in return it should be given some PSP22 token, but PSP22 mechanics will need to ask `A` whether it is ok with receiving money

Reply:
oh, I take it back - `A` should call dex with reentrancy switched on, sorry